### PR TITLE
Unwind stats when we delete verification codes

### PR DESCRIPTION
**Pull request by whaught**
## Proposed Changes

* When we fail to SMS a code, we invalidate it by deleting the code. This change also decrements the stats

**Release Note**

```release-note
Decrement stats for unsuccessful code-issuance
```

### CODE REVIEW

File: pkg/database/vercode.go [hunk 1]
```diff
@@ -106,16 +106,27 @@ func (v *VerificationCode) BeforeSave(tx *gorm.DB) error {
 // to update statistics about usage. If the executions fail, an error is logged
 // but the transaction continues. This is called automatically by gorm.
 func (v *VerificationCode) AfterCreate(scope *gorm.Scope) {
+	v.updateStats(scope, true)
+}
+
+func (v *VerificationCode) updateStats(scope *gorm.Scope, increment bool) {
 	date := timeutils.Midnight(v.CreatedAt)
 
+	var incString string
```
Comment (mikehelmick) [inline pkg/database/vercode.go:115]: simplification

```
sign := "+"
if !increment {
  sign = "-"
}
```

And then the 1 can be embedded in the statement still

Comment (whaught) [reply pkg/database/vercode.go]: sure

**Review by sethvargo — COMMENTED** [review]
I'm pretty -1 on this. I think stats should be +1 only (e.g. we need a guard now to prevent a stat from going negative), and we advertise them as best-effort.

**Comment by mikehelmick** [comment]
> I'm pretty -1 on this. I think stats should be +1 only (e.g. we need a guard now to prevent a stat from going negative), and we advertise them as best-effort.

I didn't have a chance to go back and and see where we're dispatching the increment - seems like delaying that should mitigate this issue

**Comment by whaught** [comment]
> > I'm pretty -1 on this. I think stats should be +1 only (e.g. we need a guard now to prevent a stat from going negative), and we advertise them as best-effort.
> 
> I didn't have a chance to go back and and see where we're dispatching the increment - seems like delaying that should mitigate this issue

We could do it that way, but currently we have it on a GORM AfterCreate hook which feels like the most foolproof way of never missing the increment (hard to have fail-in-the-middle). This strategy mirrors our after the fact delete of the code itself. We only do this if RowsAffected > 1, so I don't forsee this going negative if we trust our transactions.

**Comment by sethvargo** [comment]
I think it's okay if we miss an increment - all stats are advertised as best-effort.

My main concern is this is attempting to make "codes issued" match "people notified" which I'm pretty strongly opposed to. I'd rather solve this with education on what "codes issued" means.

**Pull request closed without merging**